data_mgr.py

```diff
@@ -0,0 +1,238 @@
+        items = await gen.multi(
+            get_workflow_data(*args) for args in ws_args
+            if not ids or args[0] in ids)
+        for w_id, result in items:
+            if result is not None and result != MSG_TIMEOUT:
+                new_data[w_id] = result
+        if not ids:
+            # atomic update
+            self.data = new_data
+        else:
+            # apparently faster than update(new_data)
+            for w_id, flow in new_data.items():
+                self.data[w_id] = flow
+
+    # **
+    # Data access filters
+    # **
+    def get_workflow_msgs(self, args):
+        """Return list of workflows."""
+        result = []
+        for flow_msg in self.data.values():
+            if self._workflow_filter(flow_msg, args):
+                result.append(flow_msg)
+        return result
+
+    def get_nodes_all(self, node_type, args):
+        """Return list of nodes (defns/proxies/jobs)."""
+        w_args = {
+            'ids': args['workflows'],
+            'exids': args['exworkflows'],
+        }
+        nodes = [
+            n for k in self.get_workflow_msgs(w_args)
+            for n in getattr(k, node_type)]
+        result = []
+        for node in nodes:
+            if self._node_filter(node, args):
+                result.append(node)
+        return result
+
+    def get_nodes_id(self, node_type, args):
+        """Return protobuf node objects for given id."""
+        nat_ids = args.get('native_ids', [])
+        w_ids = []
+        for nat_id in nat_ids:
+            o_name, w_name, nid = nat_id.split('/', 2)
+            w_ids.append(f"{o_name}/{w_name}")
+        if node_type == 'proxy_nodes':
+            nodes = (
+                n for w_id in set(w_ids)
+                for n in (
+                    list(getattr(self.data[w_id], 'task_proxies', []))
+                    + list(getattr(self.data[w_id], 'family_proxies', []))))
+        else:
+            nodes = (
+                n for w_id in set(w_ids)
+                for n in getattr(self.data[w_id], node_type, []))
+        result = []
+        for node in nodes:
+            if ((not nat_ids or node.id in set(nat_ids)) and
+                    self._node_filter(node, args)):
+                result.append(node)
+        return result
+
+    def get_node_id(self, node_type, args):
+        """Return protobuf node object for given id."""
+        n_id = args.get('id')
+        o_name, w_name, remainder = n_id.split('/', 2)
+        w_id = f"{o_name}/{w_name}"
+        if node_type == 'proxy_nodes':
+            nodes = (
+                list(getattr(self.data[w_id], 'task_proxies', []))
+                + list(getattr(self.data[w_id], 'family_proxies', [])))
+        else:
+            nodes = getattr(self.data[w_id], node_type, [])
+        for node in nodes:
+            if node.id == n_id:
+                return node
+        return None
+
+    def get_edges_all(self, args):
```

Comment:
Reading the documentation, it looks like it does not get all edges. But instead, all edges for a given ID (in `args` I assume). So maybe `get_edges_by_id`? Or just `get_edges`? Though I see `get_edges_id` too... not sure what's the difference?

Reply:
All workflows.